### ISSUE TITLE
Ability to subscribe to a publisher

### ISSUE THREAD

**Issue by arangamani** [go-kit/kit#209]
I have been experimenting the load balancer/service discovery components for my project and I like the idea of the load balancer keeping track of the latest endpoints. However, I would like the ability to subscribe to a publisher and get a callback when there is a change in the service. I want to control the concurrency at which I talk to the service based on the number of available endpoints. If the service endpoints change, I would like to be able to grow or shrink my pool of worker goroutines. I would like to propose something like this:

``` go
type Publisher struct {
    sync.RWMutex
    // ...
    subscribers []chan<- bool
}

/// Subscribe adds the channel to its list of subscribers. The subscribers
// will be notified when there is a change in the service.
func (p *Publisher) Subscribe(ch chan<- bool) {
    p.Lock()
    defer p.Unlock()
    p.subscribers = append(p.subscribers, ch)
}

// notify notifies all subscribers.
func (p *Publisher) notify() {
    if len(p.subscribers) == 0 {
        return
    }
    p.RLock()
    defer p.RUnlock()
    for _, ch := range p.subscribers {
        ch <- true
    }
}
```

where the `notify` would be called whenever there is a change in the service.

**Comment by ChrisHines** [comment]
I'm not prepared to comment on whether this is a good idea or not, but I wonder if a corresponding `Unsubscribe` method is needed. It may complicate the implementation of `notify`, especially regarding  handling races between `notify` and `Unsubscribe` without deadlocking.

**Comment by peterbourgon** [comment]
> I want to control the concurrency at which I talk to the service based on the number of available endpoints. If the service endpoints change, I would like to be able to grow or shrink my pool of worker goroutines.

Why?

**Comment by arangamani** [comment]
My target service (the one behind the endpoint) is somewhat special can only accept one request at a time and I don't want the requests to be queued at the server side. This way I can trigger the target service to scale up if more requests are queued at the client side.

**Comment by peterbourgon** [comment]
If you do a bit of Git archaeology you'll find that the Publisher was originally implemented very similar to your proposal. Minor difference: rather than being a `chan bool` they were a `chan []Endpoint`, so subscribers got the latest set of endpoints as a stream, rather than having to poll the Publisher. I changed away from that model because it was unnecessarily complex to satisfy the API I wanted to provide i.e. the current semi-opaque pool of endpoints.

We can certainly re-add this functionality. You'd want to model the subscribers as a `map[chan<- bool]struct{}` so you can have an Unsubscribe, as Chris suggests. But I'm not sure how I feel about promoting this to the Publisher interface. I'm afraid your use case is relatively esoteric, and I'd hate to unnecessarily complect all of the concrete Publisher implementations. Let me sleep on it, if you don't mind.

**Comment by arangamani** [comment]
No worries, take your time. I just didn't want to poll the publisher for the number of available endpoints as consul (for example) has a nice watch method.

**Comment by peterbourgon** [comment]
For sure, if we do this we should take advantage of any watch-style methods on the underlying system.

**Comment by peterbourgon** [comment]
I can't guarantee I will accept it, but if you felt up to it, I'd be very eager to review a PR, to see what the effect would be.

**Comment by peterbourgon** [comment]
I've updated slightly sd.Subscriber (née loadbalancer.Publisher) but I still think this feature would be valuable. If someone would like to try their hand at a PR, please see me for design direction :)

**Comment by arangamani** [comment]
@peterbourgon I would be interested in it. Let me know if you have any design notes to share.

**Comment by yurishkuro** [comment]
@arangamani I think what you are after has been implemented in #492. Can you confirm?

**Comment by arangamani** [comment]
@yurishkuro Looks like it will but I haven't dug into the details. I'll close this issue.
